### PR TITLE
support multiple yaml config files 

### LINKED ISSUE #201: If there are many sites, can I include conf files like nginx?
If there are many sites
For example, thousands of sites, all written in a config.yaml is too long and ugly

Can you include conf files like nginx?
That is, include files in a directory to make it easier to distinguish

### PR DESCRIPTION
**Pull request by samanhappy**
This PR is trying to resolve #201 by merging multiple yaml files.

The merge work is completed by [https://github.com/mikefarah/yq](https://github.com/mikefarah/yq) using merge operator `*` and append arrays flag `+`, more details can refer [https://mikefarah.gitbook.io/yq/operators/multiply-merge](https://mikefarah.gitbook.io/yq/operators/multiply-merge).

### CODE REVIEW

**Comment by haoel** [comment]
I am not sure this merge function would work well.

because the array is nested in the notify section. and for settings, we should merge the settings instead of overwriting the whole settings

I think we can do this better.

how about just using the yq? here is an example use it as a library: https://github.com/kamolhasan/yq-for-strings

**Comment by samanhappy** [comment]
The settings are merged already, the example maybe a little misleading, so I modified it and will add more tests later.

The merge code is simple now, I will research the yq to see if there is a better solution.

**Comment by haoel** [comment]
At least, I think we must pass the following  mandatory test cases:

## 1. Section Merge Case

### into

```yaml
http:
  - name: "test 1"
    url: "http://localhost:8080"
    method: "GET"
tcp:
  - name: "test 2"
    host: "localhost:8080"
```

### from

```yaml
notify:
  slack:
    - name: "slack"
      webhook: "https://hooks.slack.com/services/xxxxxx"
```
### result

```yaml
http:
  - name: "test 1"
    url: "http://localhost:8080"
    method: "GET"
tcp:
  - name: "test 2"
    host: "localhost:8080"
notify:
  slack:
    - name: "slack"
      webhook: "https://hooks.slack.com/services/xxxxxx"
```

## 2. Same Probe Merge Case


### into
```yaml
http:
    - name: "test 1"
        url: "http://localhost:8080"
        method: "GET"
```
### from
```yaml
http:
    - name: "test 2"
        url: "http://localhost:8181"
        method: "GET"
```
### result
```yaml
http:
  - name: "test 1"
    url: "http://localhost:8080"
    method: "GET"
  - name: "test 2"
    url: "http://localhost:8181"
    method: "GET"
```

## 3. Notify Merge Case
### into
```yaml
notify:
  slack:
    - name: slack
      webhook: "https://hooks.slack.com/services/xxxxxx"
```
### from
```yaml
notify:
  discord:
    - name: discord
      webhook: "https://discord.com/api/webhooks/xxxxxx"
```

### result
```yaml
notify:
  slack:
    - name: slack
      webhook: "https://hooks.slack.com/services/xxxxxx"
  discord:
    - name: discord
      webhook: "https://discord.com/api/webhooks/xxxxxx"
```


## 4. Notify Array Merge Case

### into
```yaml
notify:
  slack:
    - name: slack1
      webhook: "https://hooks.slack.com/services/xxxxxx"
```
### from
```yaml
notify:
  slack:
    - name: slack1
      webhook: "https://hooks.slack.com/services/xxxxxx"
```

### result
```yaml
notify:
  slack:
    - name: slack1
      webhook: "https://hooks.slack.com/services/xxxxxx"
    - name: slack1
      webhook: "https://hooks.slack.com/services/xxxxxx"
```

## 5. Settings Merge Case

### into
```yaml
settings:
  name: easeprobe
  sla:
    schedule: "daily"
    time: "00:00"
  notify:
    retry:
      times: 5
      interval: 10
```
### from
```yaml
settings:
  name: easeprobe_from
  probe:
    timeout: 10s
    interval: 30s
   sla:
    schedule: "weekly"
```
### result
```yaml
settings:
  name: easeprobe_from
  probe:
    timeout: 10s
    interval: 30s
  notify:
    retry:
      times: 5
      interval: 10
  sla:
    schedule: "weekly"
```

File: conf/merge_test.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+func TestNotifyArrayMerge(t *testing.T) {
+	into := `
+notify:
+  slack:
+    - name: slack1
+      webhook: "https://hooks.slack.com/services/xxxxxx"`
+	from := `
+notify:
+  slack:
+    - name: slack1
```
Comment (haoel) [inline conf/merge_test.go:117]:
```suggestion
    - name: slack2
```

File: conf/merge_test.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+      webhook: "https://hooks.slack.com/services/xxxxxx"`
+	expected := `
+notify:
+  slack:
+    - name: slack1
+      webhook: "https://hooks.slack.com/services/xxxxxx"
+    - name: slack1
```
Comment (haoel) [inline conf/merge_test.go:124]:
```suggestion
    - name: slack2
```

File: conf/merge_test.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+	assertMerge(t, into, from, expected)
+}
+
+func TestSameProbeMerge(t *testing.T) {
+	into := `
+http:
+	- name: "test 1"
+		url: "http://localhost:8080"
+		method: "GET"`
```
Comment (haoel) [inline conf/merge_test.go:67]:
```suggestion
http:
  - name: "test 1"
     url: "http://localhost:8080"
     method: "GET"`
```

File: conf/merge_test.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+	from := `
+http:
+	- name: "test 2"
+		url: "http://localhost:8181"
+		method: "GET"`
```
Comment (haoel) [inline conf/merge_test.go:72]:
```suggestion
http:
  - name: "test 2"
     url: "http://localhost:8181"
     method: "GET"`
```

File: conf/merge.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+	// It's possible to handle this with a mass of reflection, but we only need
+	// to merge whole YAML files. Since we're always unmarshaling into
+	// interface{}, we only need to handle a few types. This ends up being
+	// cleaner if we just handle each case explicitly.
+	if into == nil {
+		return from, nil
+	}
+	if from == nil {
+		// Allow higher-priority YAML to explicitly nil out lower-priority entries.
+		return nil, nil
```
Comment (haoel) [inline conf/merge.go:113]: I don't understand this. it should keep the `into` without change.

File: conf/merge.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		}
+		d := yaml.NewDecoder(bytes.NewReader(b))
+
+		var contents interface{}
+		if err := d.Decode(&contents); err == io.EOF {
+			// Skip empty and comment-only sources, which we should handle
+			// differently from explicit nils.
+			continue
+		} else if err != nil {
+			return nil, fmt.Errorf("couldn't decode source: %v", err)
```
Comment (haoel) [inline conf/merge.go:78]: suggestion, log the error message and keep processing the next file.

File: conf/merge.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if len(files) == 0 {
+		return nil, fmt.Errorf("yaml file not found")
+	}
+	for _, f := range files {
+		b, err := ioutil.ReadFile(f)
+		if err != nil {
+			return nil, err
```
Comment (haoel) [inline conf/merge.go:68]: suggestion, log the error message and keep processing the next file.

File: conf/merge.go (hunk 5 above)
Comment (samanhappy) [reply conf/merge.go]: this is try to rewrite nil in from, but return into is better because merge order is not certain

File: conf/merge_test.go [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+  probe:
+    timeout: 10s
+    interval: 30s
+  notify:
+    retry:
+      times: 5
+      interval: 10
+  sla:
+    schedule: "weekly"
+		time: "00:00"`
```
Comment (haoel) [inline conf/merge_test.go:161]:
```suggestion
    schedule: "weekly"
    time: "00:00"`
```

File: conf/merge.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+		if err != nil {
+			log.Errorf("Failed to read file: %v by: %v", f, err)
+			continue
+		}
+		d := yaml.NewDecoder(bytes.NewReader(b))
+
+		var contents interface{}
+		if err := d.Decode(&contents); err == io.EOF {
+			// Skip empty and comment-only sources, which we should handle
+			// differently from explicit nils.
```
Comment (haoel) [inline conf/merge.go:77]: suggestion: log a warning or info here. most of the time, we can use a log message to replace the comment.  ;-)

File: conf/merge.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+	apath, err := filepath.Abs(path)
+	if err != nil {
+		return nil, err
+	}
+	files, err := filepath.Glob(apath + "/*.yaml")
+	if err != nil {
+		return nil, err
+	}
+	if len(files) == 0 {
+		return nil, fmt.Errorf("yaml file not found")
```
Comment (haoel) [inline conf/merge.go:64]: do you think we should treat the empty directory as a normal case instead of an error?

File: conf/merge.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+		if err != nil {
+			return nil, err // error is already descriptive enough
+		}
+		merged = pair
+	}
+
+	buf := &bytes.Buffer{}
+	if !hasContent {
+		// No sources had any content. To distinguish this from a source with just
+		// an explicit top-level null, return an empty buffer.
```
Comment (haoel) [inline conf/merge.go:95]: suggestion: log an info or debug message here.

File: conf/merge.go [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+		return nil, err
+	}
+	files, err := filepath.Glob(apath + "/*.yaml")
+	if err != nil {
+		return nil, err
+	}
+	if len(files) == 0 {
+		return nil, fmt.Errorf("yaml file not found")
+	}
+	for _, f := range files {
```
Comment (haoel) [inline conf/merge.go:66]: suggestion: log a debug information which file is processing now.

File: conf/merge.go [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+			log.Errorf("Failed to decode file: %v by: %v", f, err)
+			continue
+		}
+
+		hasContent = true
+		pair, err := merge(merged, contents)
+		if err != nil {
+			return nil, err // error is already descriptive enough
+		}
+		merged = pair
```
Comment (haoel) [inline conf/merge.go:89]: suggestion: log a debug message that shows the merged yaml

File: conf/merge.go [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+	}
+	enc := yaml.NewEncoder(buf)
+	if err := enc.Encode(merged); err != nil {
+		return nil, fmt.Errorf("couldn't re-serialize merged YAML: %v", err)
+	}
+	return buf.Bytes(), nil
+}
+
+// merge merges different types using different rules
+func merge(into, from interface{}) (interface{}, error) {
```
Comment (haoel) [inline conf/merge.go:106]: suggestion: add some debug information here to show how the merge process works. if we can log the yaml `key` here that would be better.

File: conf/merge.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+	"bytes"
+	"path/filepath"
+
+	"github.com/mikefarah/yq/v4/pkg/yqlib"
+	log "github.com/sirupsen/logrus"
+)
+
+// mergeYamlFiles merges yaml files using yq(https://github.com/mikefarah/yq)
+func mergeYamlFiles(path string) ([]byte, error) {
+	files, err := filepath.Glob(path + "/*.yaml")
```
Comment (haoel) [inline conf/merge.go:31]: for `path + "/*.yaml"`, it's better to use `filepath.Join()` this function can help to deal with the issue of cross-platform - Windows uses `\`, linux uses `/` for path separator.

File: conf/merge.go [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+		// use evaluate merge, reference https://mikefarah.gitbook.io/yq/operators/multiply-merge
+		err = yqlib.NewAllAtOnceEvaluator().EvaluateFiles(". as $item ireduce ({}; . *+ $item )", files, printer, true, decoder)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		log.Warnf("yaml files not found for %v", path)
+	}
+
+	return buf.Bytes(), nil
```
Comment (haoel) [inline conf/merge.go:50]:
```suggestion
	if len(files) <= 0 {
		return nil, fmt.Errorf("yaml files not found for %v", path)
	}
	var buf bytes.Buffer
	decoder := yqlib.NewYamlDecoder()
	encoder := yqlib.NewYamlEncoder(2, false, !false, true)
	printer := yqlib.NewPrinter(encoder, yqlib.NewSinglePrinterWriter(bufio.NewWriter(&buf)))
	// use evaluate merge, reference https://mikefarah.gitbook.io/yq/operators/multiply-merge
	err = yqlib.NewAllAtOnceEvaluator().EvaluateFiles(". as $item ireduce ({}; . *+ $item )", files, printer, true, decoder)
	if err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
```

**Review by haoel — APPROVED** [review]
LGTM

**Pull request merged**